mlx_lm/awq.py

```diff
@@ -0,0 +1,541 @@
+                        module.indices = mx.concatenate(
+                            [module.indices, indices], axis=0
+                        )
+                    else:
+                        module.indices = indices
+
+                return module(x, *args, **kwargs)
+
+        return Catcher()
+
+    for block in tqdm(model.model.layers):
+        # Capture the input features for each of the layers in the transformer block
+        orig_leaves = block.leaf_modules()
+        capture_leaves = tree_map(capture, orig_leaves, is_leaf=nn.Module.is_module)
+        block.update_modules(capture_leaves)
+        outputs = run_layer(block, inputs, mask=mask)
+        block.update_modules(orig_leaves)
+        del capture_leaves
+
+        # Quantize the block without AWQ to obtain a reference loss
+        nn.quantize(block, group_size=group_size, bits=bits)
+        outputs_q = run_layer(block, inputs, mask=mask)
+        before_loss = mse(outputs, outputs_q).sum()
+        if group is not None:
+            before_loss = (
+                mx.distributed.all_sum(before_loss, stream=mx.cpu) / group.size()
+            )
+        before_loss /= outputs.size
+        block.update_modules(orig_leaves)
+        del orig_leaves
+
+        scale_block(
+            block=block,
+            mask=mask,
+            quantize_func=quantize_func,
+            n_grid=n_grid,
+        )
+
+        clip_block(
+            block=block,
+            quantize_func=quantize_func,
+            group_size=group_size,
+            n_grid=n_grid,
+        )
+
+        # Quantize the scaled and clipped block
+        nn.quantize(block, group_size=group_size, bits=bits)
+        outputs_q = run_layer(block, inputs, mask=mask)
+        after_loss = mse(outputs, outputs_q).sum()
+        if group is not None:
+            after_loss = (
+                mx.distributed.all_sum(after_loss, stream=mx.cpu) / group.size()
+            )
+        after_loss /= outputs.size
+        tqdm.write(f"Loss reduction: {after_loss / before_loss}")
+
+        inputs = outputs
+
+        mx.eval(block)
+        mx.metal.clear_cache()
+
+    if hasattr(model, "lm_head"):
+        model.lm_head = model.lm_head.to_quantized(
+            group_size=embed_group_size, bits=embed_bits
+        )
+
+
+def load_wikitext(
+    tokenizer, num_samples: int = 32, sequence_length: int = 2048, split: str = "train"
+) -> mx.array:
+    dataset = load_dataset("Salesforce/wikitext", "wikitext-2-raw-v1", split=split)
+    texts = "\n\n".join(dataset["text"])
+    tokens = tokenizer.encode(texts, return_tensors="mlx")[0]
+
+    # Select random chunks
+    starts = mx.random.randint(
+        0, len(tokens) - sequence_length - 1, shape=(num_samples, 1)
+    )
+    data = tokens[starts + mx.arange(sequence_length)]
+    if tokenizer.bos_token_id:
+        data = mx.concatenate(
+            [mx.full((*data.shape[:1], 1), tokenizer.bos_token_id), data], axis=-1
+        )
```

Comment:
It's a bit odd to use the bos token here since the segments are just randomly pulled out from a long stretch of text. I'm not sure if you tried without it.. but my guess is it's probably not useful and maybe even slightly worse to use it. I might just use `sequence_length+1` segments instead.